File: src/main/java/com/adinsight/agent/tools/ToolPermission.java
package com.adinsight.agent.tools;

public record ToolPermission(Behavior behavior, String message) {
    public enum Behavior {
        ALLOW,
        DENY
    }

    public static ToolPermission allow() {
        return new ToolPermission(Behavior.ALLOW, null);
    }

    public static ToolPermission deny(String message) {
        return new ToolPermission(Behavior.DENY, message);
    }

    public boolean denied() {
        return behavior == Behavior.DENY;
    }
}

File: src/tools/tool-permission.ts
export type ToolPermission = Readonly<
  | {
      behavior: "allow";
    }
  | {
      behavior: "deny";
      message: string;
    }
>;

export function allowTool(): ToolPermission {
  return { behavior: "allow" };
}

export function denyTool(message: string): ToolPermission {
  return { behavior: "deny", message };
}

export function isToolDenied(
  permission: ToolPermission
): permission is Extract<ToolPermission, { behavior: "deny" }> {
  return permission.behavior === "deny";
}
